homer0/projext
# TypeScript support

### What does this PR do?

This PR contains the changes to support TypeScript on the settings, so the bundle engines can use, and to support TypeScript Node targets that don't use bundling.

Here's a detailed list of all the things that were done:

#### Object rest and spread as a Babel "feature"

One of the requirements in order to transpile TypeScript with Babel is to use the [`@babel/plugin-proposal-object-rest-spread`](https://yarnpkg.com/en/package/@babel/plugin-proposal-object-rest-spread) plugin for object rest and spread, so it was added to the features you can enable/disable for your target Babel settings. The key name is `objectRestSpread`.

#### Babel configuration for TypeScript targets

The `BabelConfiguration` service now checks if the target uses TypeScript (by evaluating the `typeScript` setting) adds [`@babel/preset-typescript`](https://yarnpkg.com/en/package/@babel/preset-typescript) and, in case they are not already enabled, class properties and object read & spread support.

#### TypeScript detection on the targets finder

The `TargetsFinder` service is now able to identify `.ts` and `.tsx` extensions in order to flag a target for TypeScript. If the extension is `.tsx`, it even adds `react` as framework (if no other was detected).

The service also enables source maps for both development and production environments.

#### Targets with `.ts` and `.tsx` are identified as TypeScripts targets when loaded

Added to what the `TargetsFinder` service does and in order to add more automatic detections (in case the finder is disabled), the `Targets` service also checks the extensions of the entry files to set the `typeScript` and `framework` settings.

#### Build transpiler now transpiles from `.tsx?` to `.js`

When building/running a TypeScript Node target, the `BuildTranspiler` service takes care of correctly transpiling the files and rename them to `.js`.

#### Node targets now have support for source maps.

Well, at the beginning, having source maps for Node targets didn't make a lot of sense as they weren't being uglified and the loss with transpilation wasn't that big, but with TypeScript, it became really necessary to have them.

Just like browser targets, Node targets now have the following setting:

```js
sourceMap: {
  development: false,
  production: true,
}
```

> The source map is only generated when the target is bundled or transpiled.

Now, for bundled target, there's no issue, the engine will take care of generating the map, but for transpiled target... that needed to be done on the `BuildTranspiler` service, which brings to the **breaking change** of this PR:

The 3 transpilation methods of `BuildTranspiler` have changed their signatures in order to support a new `buildType` parameter, which will be used to determine if a source map needs to be generated for the file or not.

```diff
- transpileTargetFiles(target) { ... }
+ transpileTargetFiles(target, buildType = 'development') { ... }

- transpileFile(filepath, options = null, writeFile = true) { ... }
+ transpileFile(filepath, buildType = 'development', options = null, writeFile = true) { ... }

- transpileFileSync(filepath, options = null, writeFile = true) { ... }
+ transpileFileSync(filepath, buildType = 'development', options = null, writeFile = true) { ... }
```

#### Add a build step for TypeScript declarations

One of the things I didn't have in mind when I built the feature was the TypeScript declaration files: If you are going to distribute your project, you would probably want to distribute its types too.

The problem with this was that you can't generate the declarations with Babel, you need to do it with the TypeScript binary. I added a `BuildTypeScriptHelper` service, for this and future interactions with the TypeScript binary, that's called from the CLI build command in order to add the declarations generation to the build steps of a TypeScript target.

The declarations are only generated when the target is being "built", not when the target is executed nor when the target is being watched. And since the build steps include the bundle engine call, the same rules applies for bundled targets.

### How should it be tested manually?

1. Try a target with just an index.ts and see if projext detects it, builds it and runs it.
2. Validate the generated source map.
3. Validate the declarations.

And...

```bash
yarn test
# or
npm test
```
